File: apps/mobile/lib/core/utils/package_script_run.dart
List<String> buildRunCommand(
  String packageManager,
  String script,
  String? args,
) {
  final trimmedArgs = args?.trim();
  final tokens = trimmedArgs != null && trimmedArgs.isNotEmpty
      ? trimmedArgs.split(RegExp(r'\s+')).where((token) => token.isNotEmpty).toList()
      : <String>[];
  switch (packageManager) {
    case 'bun':
      return tokens.isNotEmpty
          ? ['bun', 'run', script, ...tokens]
          : ['bun', 'run', script];
    case 'pnpm':
      if (tokens.isEmpty) {
        return ['pnpm', 'run', script];
      }
      return tokens.first == '--'
          ? ['pnpm', 'run', script, ...tokens]
          : ['pnpm', 'run', script, '--', ...tokens];
    case 'yarn':
      if (tokens.isEmpty) {
        return ['yarn', 'run', script];
      }
      return tokens.first == '--'
          ? ['yarn', 'run', script, ...tokens]
          : ['yarn', 'run', script, '--', ...tokens];
    default:
      if (tokens.isEmpty) {
        return ['npm', 'run', script];
      }
      return tokens.first == '--'
          ? ['npm', 'run', script, ...tokens]
          : ['npm', 'run', script, '--', ...tokens];
  }
}

String formatRunCommand(
  String packageManager,
  String script,
  String? args,
) {
  return buildRunCommand(packageManager, script, args).join(' ');
}

File: apps/desktop/src/shared/package-script-run.ts
import type { PackageManagerKind } from "./ipc";

export function buildRunCommand(packageManager: PackageManagerKind, script: string, args?: string): string[] {
  const trimmedArgs = args?.trim();
  const tokens = trimmedArgs ? splitShellArgs(trimmedArgs) : [];
  switch (packageManager) {
    case "bun":
      return tokens.length > 0 ? ["bun", "run", script, ...tokens] : ["bun", "run", script];
    case "pnpm":
      return tokens.length > 0
        ? tokens[0] === "--"
          ? ["pnpm", "run", script, ...tokens]
          : ["pnpm", "run", script, "--", ...tokens]
        : ["pnpm", "run", script];
    case "yarn":
      return tokens.length > 0
        ? tokens[0] === "--"
          ? ["yarn", "run", script, ...tokens]
          : ["yarn", "run", script, "--", ...tokens]
        : ["yarn", "run", script];
    default:
      return tokens.length > 0
        ? tokens[0] === "--"
          ? ["npm", "run", script, ...tokens]
          : ["npm", "run", script, "--", ...tokens]
        : ["npm", "run", script];
  }
}

export function formatRunCommand(packageManager: PackageManagerKind, script: string, args?: string): string {
  return buildRunCommand(packageManager, script, args).join(" ");
}

function splitShellArgs(value: string): string[] {
  return value.split(/\s+/).filter(Boolean);
}
